// @semaphore-protocol/group/dist/index.node.js
/**
 * @module @semaphore-protocol/group
 * @version 3.1.0
 * @file A library to create and manage Semaphore groups.
 * @copyright Ethereum Foundation 2022
 * @license MIT
 * @see [Github]{@link https://github.com/semaphore-protocol/semaphore/tree/main/packages/group}
*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var incrementalMerkleTree = require('@zk-kit/incremental-merkle-tree');
var poseidon = require('poseidon-lite');
var bignumber = require('@ethersproject/bignumber');
var bytes = require('@ethersproject/bytes');
var keccak256 = require('@ethersproject/keccak256');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var poseidon__default = /*#__PURE__*/_interopDefaultLegacy(poseidon);

/**
 * Creates a keccak256 hash of a message compatible with the SNARK scalar modulus.
 * @param message The message to be hashed.
 * @returns The message digest.
 */
function hash(message) {
    message = bignumber.BigNumber.from(message).toTwos(256).toHexString();
    message = bytes.zeroPad(message, 32);
    return BigInt(keccak256.keccak256(message)) >> BigInt(8);
}

var Group = /** @class */ (function () {
    /**
     * Initializes the group with the group id and the tree depth.
     * @param id Group identifier.
     * @param treeDepth Tree depth.
     */
    function Group(id, treeDepth) {
        if (treeDepth === void 0) { treeDepth = 20; }
        if (treeDepth < 16 || treeDepth > 32) {
            throw new Error("The tree depth must be between 16 and 32");
        }
        this._id = id;
        this.merkleTree = new incrementalMerkleTree.IncrementalMerkleTree(poseidon__default["default"], treeDepth, hash(id), 2);
    }
    Object.defineProperty(Group.prototype, "id", {
        /**
         * Returns the id of the group.
         * @returns Group id.
         */
        get: function () {
            return this._id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Group.prototype, "root", {
        /**
         * Returns the root hash of the tree.
         * @returns Root hash.
         */
        get: function () {
            return this.merkleTree.root;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Group.prototype, "depth", {
        /**
         * Returns the depth of the tree.
         * @returns Tree depth.
         */
        get: function () {
            return this.merkleTree.depth;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Group.prototype, "zeroValue", {
        /**
         * Returns the zero value of the tree.
         * @returns Tree zero value.
         */
        get: function () {
            return this.merkleTree.zeroes[0];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Group.prototype, "members", {
        /**
         * Returns the members (i.e. identity commitments) of the group.
         * @returns List of members.
         */
        get: function () {
            return this.merkleTree.leaves;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Returns the index of a member. If the member does not exist it returns -1.
     * @param member Group member.
     * @returns Index of the member.
     */
    Group.prototype.indexOf = function (member) {
        return this.merkleTree.indexOf(member);
    };
    /**
     * Adds a new member to the group.
     * @param member New member.
     */
    Group.prototype.addMember = function (member) {
        this.merkleTree.insert(BigInt(member));
    };
    /**
     * Adds new members to the group.
     * @param members New members.
     */
    Group.prototype.addMembers = function (members) {
        for (var _i = 0, members_1 = members; _i < members_1.length; _i++) {
            var member = members_1[_i];
            this.addMember(member);
        }
    };
    /**
     * Updates a member in the group.
     * @param index Index of the member to be updated.
     * @param member New member value.
     */
    Group.prototype.updateMember = function (index, member) {
        this.merkleTree.update(index, member);
    };
    /**
     * Removes a member from the group.
     * @param index Index of the member to be removed.
     */
    Group.prototype.removeMember = function (index) {
        this.merkleTree.delete(index);
    };
    /**
     * Creates a proof of membership.
     * @param index Index of the proof's member.
     * @returns Proof object.
     */
    Group.prototype.generateMerkleProof = function (index) {
        var merkleProof = this.merkleTree.createProof(index);
        merkleProof.siblings = merkleProof.siblings.map(function (s) { return s[0]; });
        return merkleProof;
    };
    return Group;
}());

exports.Group = Group;
